apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java

```diff
@@ -137,6 +139,44 @@ private void startCaptureBody(Transaction transaction, String method, @Nullable
         }
     }
 
+    @Nullable
+    public static String getUserFromPrincipal(@Nullable Principal principal) {
+        if (principal == null) {
+            return null;
+        }
+
+        String userName;
+        if (principal instanceof Map) {
+            // Microsoft/Azure SSO fallback
+            // https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens
+
+            Map<?, ?> map = ((Map<?, ?>) principal);
+
+            userName = getFirstClaim(map, "preferred_username");
+            if (userName == null) {
+                userName = getFirstClaim(map, "name");
+            }
+        } else {
+            userName = principal.getName();
+        }
+
+        return userName;
+    }
+
+    @Nullable
+    private static String getFirstClaim(Map<?, ?> map, String key) {
+        // entries are stored in nested collection, even when there is a single entry
+        // https://docs.microsoft.com/en-us/azure/app-service/configure-language-java?pivots=platform-linux#tomcat-1
+        Object entry = map.get(key);
+        if (entry instanceof List) {
+            Object first = ((List<?>) entry).get(0);
+            if (first instanceof String) {
+                return (String) first;
+            }
```

Comment:
[minor] if we find a value for the looked-up key, don't we want to always use it?
```suggestion
            if (first != null) {
                first.toString();
            }
```